### PR TITLE
Add camera component

### PR DESCRIPTION
**Pull request by stonebuzz**
Signed-off-by: Stanislas <skita@teclib.com>

Add camera component
Camera is defined by 

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #number

### CODE REVIEW

File: inc/define.php [hunk 1]
```diff
@@ -291,6 +291,10 @@
                                                   'DeviceCase', 'DevicePowerSupply', 'DeviceGeneric',
                                                   'DeviceSimcard', 'DeviceSensor'];
 
+if (version_compare(GLPI_VERSION, '9.5.1', '>')) {
```
Comment (trasher) [inline inc/define.php:294]: Maybe check if class exists rather than compare to version

File: install/mysql/glpi-empty.sql [hunk 2]
```diff
@@ -1871,6 +1871,104 @@ CREATE TABLE `glpi_deviceharddrives` (
...
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `unicity` (`name`),
+  KEY `date_mod` (`date_mod`),
+  KEY `date_creation` (`date_creation`)
+) ENGINE=InnoDB  DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
+
+
+### Dump table glpi_resolutionvideos
+
+DROP TABLE IF EXISTS `glpi_resolutionvideos`;
```
Comment (trasher) [inline install/mysql/glpi-empty.sql:1958]: to follow the same scheme as image, maybe `glpi_videoresolutions`?

File: inc/define.php (hunk 1 above)
Comment (stonebuzz) [reply inc/define.php]: done

File: install/mysql/glpi-empty.sql (hunk 2 above)
Comment (stonebuzz) [reply install/mysql/glpi-empty.sql]: done

File: inc/resolution.class.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+class Resolution extends CommonDropdown {
+
+   public $can_be_translated = false;
+
+   static function getTypeName($nb = 0) {
+      return _n('Resolution', 'Resolutions', $nb);
```
Comment (trasher) [inline inc/resolution.class.php:42]:
```suggestion
      return _nx('video', 'Resolution', 'Resolutions', $nb);
```

Comment (trasher) [reply inc/resolution.class.php:42]: So we can keep `__('Resolution')` as is, it's not the same term definition.

File: inc/item_devicecamera_format.class.php [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+**/
+class Item_DeviceCamera_Format extends CommonDBRelation {
+
+   static public $itemtype_1 = 'Item_DeviceCamera';
+   static public $items_id_1 = 'item_devicecameras_id';
+   //static public $take_entity_1 = false;
```
Comment (orthagh) [inline inc/item_devicecamera_format.class.php:71]: comments to removed ?

**Review by orthagh — APPROVED** [review]

File: install/update_92_93.php [hunk 5]
```diff
@@ -728,6 +728,10 @@ function update92to93() {
    /** Add Item_Device* display preferences */
    $itemDeviceTypes = Item_Devices::getDeviceTypes();
    foreach ($itemDeviceTypes as $itemDeviceType) {
+
+      if (!class_exists($itemDeviceType)) {
+         continue;
+      }
```
Comment (cedric-anne) [inline install/update_92_93.php:734]: I guess `Item_Devices::getDeviceTypes()` should not return types that have no corresponding class.
Why is it necessary ?

File: install/update_95_xx/cameras.php [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      PRIMARY KEY (`id`),
+      UNIQUE KEY `unicity` (`manufacturers_id`,`devicecameramodels_id`),
+      KEY `designation` (`designation`),
+      KEY `devicecameramodels_id` (`devicecameramodels_id`),
+      KEY `entities_id` (`entities_id`),
+      KEY `is_recursive` (`is_recursive`),
+      KEY `date_mod` (`date_mod`),
+      KEY `date_creation` (`date_creation`)
+      ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "9.5 add table glpi_devicecameras");
```
Comment (cedric-anne) [inline install/update_95_xx/cameras.php:62]:
```suggestion
   $DB->queryOrDie($query, "x.x add table glpi_devicecameras");
```

File: install/update_95_xx/cameras.php [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+   $query = "CREATE TABLE `glpi_devicecameramodels` (
+      `id` int NOT NULL AUTO_INCREMENT,
+      `name` varchar(255) DEFAULT NULL,
+      `comment` text,
+      `product_number` varchar(255) DEFAULT NULL,
+      PRIMARY KEY (`id`),
+      KEY `name` (`name`),
+      KEY `product_number` (`product_number`)
+      ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "9.5 add table glpi_devicecameramodels");
```
Comment (cedric-anne) [inline install/update_95_xx/cameras.php:75]:
```suggestion
   $DB->queryOrDie($query, "x.x add table glpi_devicecameramodels");
```

File: install/update_95_xx/cameras.php [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+      `entities_id` int NOT NULL DEFAULT '0',
+      `is_recursive` tinyint NOT NULL DEFAULT '0',
+      PRIMARY KEY (`id`),
+      UNIQUE KEY `unicity` (`name`),
+      KEY `date_mod` (`date_mod`),
+      KEY `entities_id` (`entities_id`),
+      KEY `is_recursive` (`is_recursive`),
+      KEY `date_creation` (`date_creation`)
+      ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "9.5 add table glpi_imageformats");
```
Comment (cedric-anne) [inline install/update_95_xx/cameras.php:94]:
```suggestion
   $DB->queryOrDie($query, "x.x add table glpi_imageformats");
```

File: install/update_95_xx/cameras.php [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+      `is_recursive` tinyint NOT NULL DEFAULT '0',
+      PRIMARY KEY (`id`),
+      UNIQUE KEY `unicity` (`name`),
+      KEY `date_mod` (`date_mod`),
+      KEY `is_video` (`is_video`),
+      KEY `entities_id` (`entities_id`),
+      KEY `is_recursive` (`is_recursive`),
+      KEY `date_creation` (`date_creation`)
+      ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "9.5 add table glpi_resolutions");
```
Comment (cedric-anne) [inline install/update_95_xx/cameras.php:115]:
```suggestion
   $DB->queryOrDie($query, "x.x add table glpi_resolutions");
```

File: install/update_95_xx/cameras.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+      PRIMARY KEY (`id`),
+      KEY `items_id` (`items_id`),
+      KEY `devicecameras_id` (`devicecameras_id`),
+      KEY `is_deleted` (`is_deleted`),
+      KEY `is_dynamic` (`is_dynamic`),
+      KEY `entities_id` (`entities_id`),
+      KEY `is_recursive` (`is_recursive`),
+      KEY `item` (`itemtype`,`items_id`)
+      ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "9.5 add table glpi_items_devicecameras");
```
Comment (cedric-anne) [inline install/update_95_xx/cameras.php:137]:
```suggestion
   $DB->queryOrDie($query, "x.x add table glpi_items_devicecameras");
```

File: install/update_95_xx/cameras.php [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+      `id` int NOT NULL AUTO_INCREMENT,
+      `item_devicecameras_id` int NOT NULL DEFAULT '0',
+      `imageformats_id` int NOT NULL DEFAULT '0',
+      `is_dynamic` tinyint NOT NULL DEFAULT '0',
+      PRIMARY KEY (`id`),
+      KEY `item_devicecameras_id` (`item_devicecameras_id`),
+      KEY `imageformats_id` (`imageformats_id`),
+      KEY `is_dynamic` (`is_dynamic`)
+   ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "9.5 add table glpi_items_devicecameras_formats");
```
Comment (cedric-anne) [inline install/update_95_xx/cameras.php:151]:
```suggestion
   $DB->queryOrDie($query, "x.x add table glpi_items_devicecameras_formats");
```

File: install/update_95_xx/cameras.php [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+      `id` int NOT NULL AUTO_INCREMENT,
+      `item_devicecameras_id` int NOT NULL DEFAULT '0',
+      `resolutions_id` int NOT NULL DEFAULT '0',
+      `is_dynamic` tinyint NOT NULL DEFAULT '0',
+      PRIMARY KEY (`id`),
+      KEY `item_devicecameras_id` (`item_devicecameras_id`),
+      KEY `resolutions_id` (`resolutions_id`),
+      KEY `is_dynamic` (`is_dynamic`)
+   ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "9.5 add table glpi_items_devicecameras_resolutions");
```
Comment (cedric-anne) [inline install/update_95_xx/cameras.php:165]:
```suggestion
   $DB->queryOrDie($query, "x.x add table glpi_items_devicecameras_resolutions");
```

File: install/update_92_93.php (hunk 5 above)
Comment (trasher) [reply install/update_92_93.php:734]: No, as far as I've seen, class existance is not checked before

File: inc/inventory/asset/device.class.php [hunk 13]
```diff
@@ -99,6 +99,10 @@ public function handle() {
                   'is_dynamic'         => 1,
                ] + (array)$val;
                $itemdevice->add($itemdevice_data, [], $this->withHistory());
+
+               if (method_exists($this, 'itemdeviceAdded')) {
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/inventory/asset/device.class.php]: Done

File: inc/item_devicecamera_format.class.php [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+            self::getType(),
+            //TRANS : %1$s is the itemtype name,
+            //        %2$s is the name of the item (used for headings of a list)
+            sprintf(
+               __('%1$s = %2$s'),
+               $camera->getTypeName(1),
+               $camera->getName()
+            )
+         );
+      }
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/item_devicecamera_format.class.php]: Done

File: inc/item_devicecamera_format.class.php [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+ * GLPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+**/
+class Item_DeviceCamera_Format extends CommonDBRelation {
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/item_devicecamera_format.class.php]: Done

File: inc/item_devicecamera_resolution.class.php [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+            self::getType(),
+            //TRANS : %1$s is the itemtype name,
+            //        %2$s is the name of the item (used for headings of a list)
+            sprintf(
+               __('%1$s = %2$s'),
+               $camera->getTypeName(1),
+               $camera->getName()
+            )
+         );
+      }
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/item_devicecamera_resolution.class.php]: Done

File: install/update_92_93.php (hunk 5 above)
Comment (cedric-anne) [reply install/update_92_93.php:734]: Indeed, but why this check is necessary ?
If a device return a bad value on its `getItem_DeviceType` method (which is the method used to populate `Item_Devices::getDeviceTypes()` result, then maybe it should be fixed there.

Comment (trasher) [reply install/update_92_93.php]: I've removed; all seems to be ok

Comment (cedric-anne) [reply install/update_92_93.php]: :+1:

**Review by cedric-anne — APPROVED** [review]

File: inc/inventory/asset/camera.class.php [hunk 17]
```diff
@@ -79,21 +80,37 @@ protected function itemdeviceAdded(Item_Devices $itemdevice, $val) {
...
       }
 
-      $cam_resolutions = new \Item_DeviceCamera_ImageResolution();
-      $cam_resolutions->add([
-         'item_devicecameras_id' => $itemdevice->fields['id'],
-         'imageresolutions_id' => $resolution->fields['id'],
-         'is_dynamic' => 1
-      ]);
+      foreach ($val as $rsl) {
+         if (empty($val)) {
```
Comment (cedric-anne) [inline inc/inventory/asset/camera.class.php:88]: I guess you used wrong var.
```suggestion
         if (empty($rsl)) {
```

**Comment by stonebuzz** [comment]
TU failed because of inventory format
need this -> https://github.com/glpi-project/inventory_format/pull/36

**Pull request merged**